cpp/daal/src/externals/service_rng_ref.h

```diff
@@ -0,0 +1,363 @@
+        if (!std::is_floating_point<T>::value) return services::ErrorDataTypeNotSupported;
+        std::uniform_real_distribution<double> distrib(static_cast<double>(a), static_cast<double>(b));
+        for (size_t i = 0; i < n; i++)
+        {
+            r[i] = static_cast<T>(distrib(_gen));
+        }
+        _nSkip += n;
+        return errcode;
+    }
+    template <typename T>
+    int fGaussian(const size_t n, T * r, const T a, const T sigma, const int method)
+    {
+        int errcode = 0;
+        if (rngSeedSize() == 0 || rngSeed() == nullptr) return services::ErrorNullInput;
+        if (!std::is_floating_point<T>::value) return services::ErrorDataTypeNotSupported;
+        std::normal_distribution<T> distrib(a, sigma);
+        for (size_t i = 0; i < n; i++)
+        {
+            r[i] = distrib(_gen);
+        }
+        _nSkip += n;
+        return errcode;
+    }
+    template <typename T>
+    int iBernoulli(const size_t n, T * r, const double p, const int method)
+    {
+        int errcode = 0;
+        if (rngSeedSize() == 0 || rngSeed() == nullptr) return services::ErrorNullInput;
+        if (!std::is_integral<T>::value) return services::ErrorDataTypeNotSupported;
+        std::bernoulli_distribution distrib(p);
+        for (size_t i = 0; i < n; i++)
+        {
+            r[i] = distrib(_gen);
+        }
+        _nSkip += n;
+        return errcode;
+    }
+
+private:
+    Gen _gen;
+    int _brngId;
+    size_t _seedSize;
+    unsigned int * _seed;
+    size_t _nSkip;
+};
+
+template <CpuType cpu>
+class BaseRNG : public BaseRNGIface<cpu>
+{
+public:
+    BaseRNG(const unsigned int seed, const int brngId) : _stream(0)
+    {
+        unsigned int tempSeed = seed;
+        if (genTypeIsSame(brngId, __DAAL_BRNG_MT2203))
+        {
+            // TODO Replace workaround
+            tempSeed += brngId - __DAAL_BRNG_MT2203;
+        }
+        _stream = new State<>(brngId, 1, &tempSeed);
+    }
+
+    BaseRNG(const size_t n, const unsigned int * seed, const int brngId = __DAAL_BRNG_MT19937) : _stream(0)
+    {
+        _stream = new State<>(brngId, n, seed);
+    }
+
+    BaseRNG(const BaseRNG<cpu> & other) : _stream(0) { _stream = other._stream->clone(); }
+
+    ~BaseRNG()
+    {
+        if (_stream != nullptr)
+        {
+            delete _stream;
+            _stream = nullptr;
+        }
+    }
+
+    int getStateSize() const { return _stream->getSize(); }
+
+    int saveState(void * dest) const
+    {
+        int errcode = 0;
+        _stream->clone((char *)dest);
+        return errcode;
+    }
+
+    int loadState(const void * src)
+    {
+        int errcode = 0;
+        _stream->assign(src);
+        return errcode;
+    }
+
+    // Not implemented
+    int leapfrog(size_t threadNum, size_t nThreads)
+    {
+        int errcode = 0;
+        // TODO add call for (vslLeapfrogStream, (_stream, threadNum, nThreads), errcode);
```

Comment:
It is needed either to report an error or to make some implementation.

Reply:
oneDAL implementation of the backend (externals) does not provide any capabilities for error reporting, as I can see.

Reply:
I think it would be enough to return `ErrorLeapfrogUnsupported` status code instead of zero.